### PR TITLE
EG: Add Benchmark Notebook For Annotations

### PR DESCRIPTION
**Pull request by John-P**
This is a work in progress but sharing as a PR now to get early feedback.

### CODE REVIEW

**Review by DavidBAEpstein — COMMENTED** [review]
Awaiting non-draft version. Nothing to add.

**Comment by John-P** [comment]
I've significantly updated this now and probably ready for review comments. I have also found and fixed some bugs in the process of making this notebook. I have so far only profiled run time. I am looking into adding a memory consumption at the end if we have time before submitting responses.

File: benchmarks/annotation_store.ipynb [hunk 1]
```diff
@@ -0,0 +1,2103 @@
...
...zh27BgWLVpU1WUQEVUrfAhTBanuz9OgInyeRuXhQ5gqT1kPYaq2PQ0iIip/DA0iIpKMoUFERJIxNIiISDKGBhERScbQICIiyRgaREQkGUODiIgkY2gQEZFkDA0iIpKMoUFERJIxNIiISDKGBhERScbQICIiyaptaPB5GkRE5a/aPrkvPDwc4eHhVV0GEVG1Um17GkREVP4YGkREJBlDg4iIJGNoEBGRZAwNIiKSjKFBRESSMTSIiEgyhgYREUnG0CAiIskYGkREJBlDg4iIJGNoEBGRZAwNIiKSjKFBRESSMTSIiEgyhgYREUnG0CAiIskYGkREJBlDg4iIJKu2oXHs2DEsWrSoqssgIqpWlFVdQEUJDw9HeHh4VZdBRFStVNueBhERlT+GBhERScbQICIiyRgaREQkGUODiIgkY2gQEZFkDA0iIpKMoUFERJIxNIiISDKGBhERScbQICIiyRgaREQkGUODiIgkY2gQEZFkDA0iIpKMoUFERJIxNIiISLIyn9yn1+uxd+9exMfH488//4TBYICDgwOCgoLQrFkztGvXDi4uLpVVKxERVbFSQ2Pt2rXYt28fnn76aXTo0AEBAQHQ6XQwGo1ISUlBYmIiJk6ciNatW6N///6VWTMREVWRUkPD3d0dc+fOhUqlshsXEhKC1q1bo6CgALt27arQAomI6NFRamh07dr1vhOr1Wp06dKlXAsiIqJHV5nHNIqdPn0aPj4+8PHxwe3bt/Hdd99BLpejX79+cHNzq+ASiYjoUSHp7Klly5ZBLi9qunr1apjNZshkMixatKhCiyMiokeLpJ5GZmYmvLy8YDab8ccff+Drr7+GUqnEsGHDKro+IiJ6hEgKDZ1Oh6ysLCQnJyMwMBBarRYmkwkmk6mi6yMiokeIpNDo0qULJk2aBJPJhEGDBgEAzp49i4CAgIqsjYiIHjGSQuOVV15BixYtIJfLUaNGDQCAh4cHhg8fXqHFERHRo0VSaACAv79/ma+JiKj6K/XsqUmTJuHQoUOlHrcwmUw4ePAgJk+eXGHF/R3Hjh3j2V1EROVMJoQQJY24du0a1q9fj8TERISEhMDf3x9arRZ5eXlITU3FpUuX0KhRI7z66qsIDAys7LofyPXr1yt9meYh3Sp9mVT5FEu2VHUJTwwvLy9kZGRUdRlPhLL2JJUaGsWysrJw8uRJXL16Fbm5u
```
Comment (DavidBAEpstein) [inline benchmarks/annotation_store.ipynb:1378]: 1) Is there a replacement for wget that is part of standard bash on the usual platforms?
2) Could the error message be more enlightening? Something like 
On Macos: EITHER brew install wget OR sudo port install wget
Perhaps that would already be obvious to users.

**Review by DavidBAEpstein — APPROVED** [review]
benchmarks/annotation_store.ipynb: nice demo and interesting comparisons and explanations
tiatoolbox/annotation/storage.py: in line 666 of the current version in your branch,
replace "string which define" by "string defining"

File: benchmarks/annotation_store.ipynb (hunk 1 above)
Comment (John-P) [reply benchmarks/annotation_store.ipynb:1378]: We could possibly swap to curl which is more commonly pre-installed or use requests (which tiatoolbox installs as a dependency) and do it in python.

File: benchmarks/annotation_store.ipynb [hunk 2]
```diff
@@ -0,0 +1,2103 @@
+{
```
Comment (shaneahmed) [inline benchmarks/annotation_store.ipynb:1]: <p>Please can we add TIAToolbox notebooks preamble. For example, link to open in colab/kaggle and some standard description at the top?</p><p> </p><p>I would also suggest to write a summary of why you have created this benchmark notebook. Probably copy some text from the manuscript.</p><p> </p><p>Also is it possible to measure these tests on colab and report?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TissueImageAnalytics/tiatoolbox/pull/320/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (shaneahmed) [inline benchmarks/annotation_store.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #9.        """
</pre><p>Add new line.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TissueImageAnalytics/tiatoolbox/pull/320/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

**Review by shaneahmed — COMMENTED** [review]
I have added some comments on the python notebook.  
Please can you also address the tests which are failing on travis?

File: benchmarks/annotation_store.ipynb (hunk 2 above)
Comment (John-P) [reply benchmarks/annotation_store.ipynb:1]: It is possible to run on colab but it's not easy to then save the output into a commit.

Comment (John-P) [reply benchmarks/annotation_store.ipynb:1]: Preamble added now (but won't work yet as it points to master).

Comment (John-P) [reply benchmarks/annotation_store.ipynb:1]: Added more introduction text and a mini tutorial for using the storage classes.

**Comment by John-P** [comment]
Looking pretty much ready now. Needs another full run through on my laptop to save the outputs before merging though I am not doing this right now because I am travelling and do not have power all the mite (it may drain the battery a lot).

File: tiatoolbox/annotation/storage.py [hunk 3]
```diff
@@ -926,7 +1080,21 @@ def clear(self) -> None:
...
+            e.g. ":memory:" or "./data.db"), a pathlib Path, or a file
+            handle.
+        path (Path):
+            The path to the annotation store data. This will be
+            ":memory:" if the annotation store is in-memory. This is
+            derived from `connection` and normalised to be a pathlib
+            Path object.
+        con (sqlite3.Connection):
+            The sqlite3 database connection.
+    """
```
Comment (shaneahmed) [inline tiatoolbox/annotation/storage.py:1097]: new line.

File: benchmarks/annotation_store.ipynb [hunk 4]
```diff
@@ -0,0 +1,2236 @@
...
```
Comment (shaneahmed) [inline benchmarks/annotation_store.ipynb:1]: <p>In the toolbox, we ...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TissueImageAnalytics/tiatoolbox/pull/320/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (shaneahmed) [inline benchmarks/annotation_store.ipynb:1]: <p>We also "," do we need a comma here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TissueImageAnalytics/tiatoolbox/pull/320/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

**Review by shaneahmed — APPROVED** [review]
Just a couple of minor comments. Once addressed we can merge.
Can you also try to improve coverage it has dropped a bit.

**Pull request merged**